### ISSUE TITLE
--add_recurrence for Knowledge Sharing Model Implementation (composition task)

### ISSUE THREAD

**Issue by Davide011** [OSU-NLP-Group/GrokkedTransformer#7]
Dear authors,
First of all, thank you for sharing this fascinating work and providing the associated code and trained weights. Your contributions have been incredibly insightful and I am currently building on your work as part of my master's thesis. I greatly appreciate your efforts in making this research accessible.

I am reaching out regarding the implementation of the --add_recurrence option in the provided code. Specifically, I am attempting to re-implement the knowledge-sharing model provided among the saved_models ("composition.2000.200.12.6_0.3_4_knowledgesharing").
However, I encountered an issue in getting a similar training result. To proceed to the implementatio I looked the config.json file of the shared model and  it seems that the only changes from the normal implementation is to add the --add_recurrence in the argparse and to change the N_layer to 4. What I have noticed after not getting any accuracy improvement for OOD casistic by looking in the provided code (seq2seq_model.py in the repository) it  seems to me at first glance that the --add_recurrence was only started to be implemented by accept the variable

"""        if add_recurrence:
            print("***in recurrence mode***")
            self.language_model.config.add_recurrence = add_recurrence ####
"""

but the implementation of what to do in the casistic add_recurrence= True is not present. Which would explain the trained model to fail on OOD in my training.

Hope you can get me some more insights and suggestions, if my issues is  related to some additional parse argumens/changes I oversaw for correctly implementing the knowledge-sharing model, or indeed the provided code didn't include the last implementation and you can share the final code implementation.

Any clarification or assistance you can provide would be immensely appreciated. Thank you again for your excellent work and for making it available to the community. 
I look forward to your response.

Best regards.

Davide

**Comment by Boshi-Wang** [comment]
Thanks for the interests!

This is implemented in ```transformers/src/transformers/models/gpt2/modeling_gpt2.py``` where the layers are simply run twice when config.add_recurrence is True.

I'm not sure why there's no OOD improvement in your case; how many training steps did you run? The OOD generalization comes very slowly so you need to train for quite some steps. Maybe you can also try a larger weight decay factor (e.g., 0.3/0.5) which should be able to speed it up.

**Comment by Davide011** [comment]
Thanks for the response! :)

Using a larger WD (I used previously only 0.01) and not noticing any increase in the expected steps shown in the appendix I stopped, and thought it was not working.

However, with your further insight, I was able to run the model and get results similar to the one shown in the appendix.

Thank you very much!

Best,
Davide
